### DISCUSSION TITLE
How to disable cache

### DISCUSSION THREAD

**Discussion by mohux** [vercel/next.js#19558] [Help]
For some weird reason, the client is asking  a really weird task to be done,

```
- Ensure the cache-control HTTP header is set with no-cache, no-store, and mustrevalidate; and that the pragma HTTP header is set with no-cache.
```

for this cache header
```
s-maxage=31536000, stale-whilerevalidate
```

how can I remove the cache headers? 

Thanks in advance

**Comment by Tomburgs** [comment]
What content are you referring to?

On API routes you can do `res.setHeader('Cache-Control', 'your-cache-policy')`.

For other pages you can refer to the [header documentation](https://nextjs.org/docs/api-reference/next.config.js/headers).

I have, however noticed that if you try to apply `Cache-Control` header to the `/_next/static/:path*` it will get overwritten with the default aggressive caching that Next.js has set which is `public, max-age=31536000, immutable`, appears to be coming from [here](https://github.com/vercel/next.js/blob/canary/packages/next/next-server/server/next-server.ts#L564) - I wonder if this is intentional @timneutkens?

**Reply by justintemps** [reply]
I believe this is breaking Preview Mode in apps that are served behind Cloudflare. The aggressive caching leads to requests with preview cookies never reaching the server. I tried to fix this by setting custom headers, but evidently Next.js overrides cache control so that this has no effect:

```
module.exports = {
 async headers() {
    return [
      {
        source: "/:path*",
        headers: [
          {
            key: "Cache-Control",
            value: "no-cache no-store",
          },
        ],
      },
    ];
  },
}
```
